File: exercises/week02/src/SmokeSensor.java
public class SmokeSensor extends HazardSensor implements BatteryPoweredSensor {
    private String location;
    private double batteryPercentage;
    private String sensorType;

    public SmokeSensor(String location) {
        this.location = location;
        this.batteryPercentage = 100.0;
        this.sensorType = "Smoke sensor";
    }

    @Override
    public boolean isTriggered() {
        if (getBatteryPercentage() != 0.0) batteryPercentage -= 20;
        return Math.random() < 0.1;
    }

    @Override
    public String getLocation() {
    return location;
  }

    @Override
    public String getSensorType() { return sensorType; }

    @Override
    public double getBatteryPercentage() {
    return batteryPercentage;
  }
}

File: exercises/week02/src/SmokeSensor.scala
import scala.util.Random

class SmokeSensor(location: String) extends HazardSensor with BatteryPowered {
  private val sensorType: String = "Smoke"
  private var batteryPercentage: Double = 100.0

  override def isTriggered: Boolean = {
    if (getBatteryPercentage != 0) batteryPercentage -= 20
    Random.nextDouble() < 0.1
  }

  override def getLocation: String = location

  override def getSensorType: String = sensorType

  override def getBatteryPercentage: Double = batteryPercentage
}
